aten/src/ATen/cuda/CUDAGenerator.cpp

```diff
@@ -1,46 +1,148 @@
-#include <ATen/Config.h>
-
 #include <ATen/CUDAGenerator.h>
-#include <ATen/Context.h>
-#include <THC/THCTensorRandom.h>
-#include <stdexcept>
+#include <c10/cuda/CUDAFunctions.h>
+
+namespace at {
 
-// There is only one CUDAGenerator instance. Calls to seed(), manualSeed(),
-// initialSeed(), and unsafeGetTH() refer to the THCGenerator on the current
-// device.
+namespace cuda { namespace detail {
 
-THCGenerator* THCRandom_getGenerator(THCState* state);
+// Ensures we only call cudaGetDeviceCount only once.
+static std::once_flag num_gpu_init_flag;
 
-namespace at {
+// Total number of gpus in the system.
+static int64_t num_gpus;
 
-CUDAGenerator::CUDAGenerator(Context * context_)
-  : Generator(Device(DeviceType::CUDA)), context(context_)
-{
+// Ensures default_gens_cuda is initialized once.
+static std::deque<std::once_flag> cuda_gens_init_flag;
+
+// Default, global CUDA generators, one per GPU.
+static std::vector<std::shared_ptr<CUDAGenerator>> default_gens_cuda;
```

Comment:
nit: Default generators have static lifetime, right? So there isn't really any reason they have to live in a `shared_ptr`. (On the other hand, this is pretty harmless; if I don't find any other issues, don't bother addressing this.)

Reply:
Oh, maybe your problem is there are mutexes in the generator and that makes them unmovable :>